ouroboros-consensus/docs/AddingAnEra.md

```diff
@@ -0,0 +1,126 @@
+
+The steps are fairly straightforward. We have put things in place (in consensus
+and in the ledger) when adding the Allegra and Mary eras that should now make it
+much easier to add a new additional era.
+
+There are two main driving changes: adding the new era to
+`Ouroboros.Consensus.Shelley.Eras` and including that era in
+`Ouroboros.Consensus.Cardano.Block`. The compiler should point the rest of the
+way. It's mostly a matter of following the existing patterns by imitating the
+previous Shelley-based case. Be sure to run both the Shelley
+(`ouroboros-consensus-shelley-test`) and the Cardano
+(`ouroboros-consensus-cardano-test`) tests.
+
+For exhaustiveness, we give an overview of the changes needed. The new era we'll
+be adding is the Alonzo era, which comes after the Mary era.
+
+## Preparation
+
+* Locate the new tag in the ledger, e.g., `AlonzoEra`. The ledger should have an
+  instance of the `ShelleyBasedEra` class (the class defined in
+  `Shelley.Spec.Ledger.API`, do not confuse it with the class with the same name
+  in consensus) for this era. This class should provide all the instances
+  consensus integration will rely on.
+
+* Note that both the affected consensus libraries and test suites are parametric
+  in the Shelley-based era or in the general era. The test suite relies on the
+  necessary `Arbitrary`, `Show`, `Eq`, ... instances to be available for all of
+  the Cardano eras. This means that a new era can only be added to the Cardano
+  eras when the ledger provides all the necessary instances for the new era,
+  *including the instances for the test suite* (e.g., the serialisation
+  generators). Note that because the test suite is parametric/generic in the
+  eras, it is not easy to temporarily omit a single era from the test suite
+  until the required instances have been added. This means that we can't extend
+  the library with the new era without extending the test suite too. In the
+  past, this has been the main blocker: the library was ready, but the required
+  instances in the ledger were missing for our test suite(s).
+
+* Add the required dependencies on the ledger package (and its test package) to
+  the `cabal.project` file. You will have to add it to some other cabal files
+  too, let the compiler tell you which ones.
+
+## `ouroboros-consensus-shelley`
+
+* Define `StandardAlonzo` in `Ouroboros.Consensus.Shelley.Eras` and add any
+  missing instances to that module, update the export list appropriately.
+
+* In `Ouroboros.Consensus.Shelley.Node`, define `ProtocolParamsAlonzo` just like
+  the existing ones. In case the era in question needs extra parameters (e.g., a
+  new genesis config?), they can be added here.
+
+## `ouroboros-consensus-shelley-test`
+
+* In `Test.Consensus.Shelley.Examples`, define an `examplesAlonzo` similar to
+  how `examplesMary` is defined. If necessary, i.e., when new type families have
+  been created, the `examples` function might have to take more arguments. This
```

Comment:
> when new type families have been created

"when new type families have been created in `cardano-ledger-specs` or `Ouroboros.Consensus.Shelley.Eras`"?

Reply:
In `cardano-ledger-specs`. For example, they recently turned `TxOut` in a type family, IIRC. I'll update the text.